File: queries/h17.R
tpc_h17 <- function(input_func, collect_func = dplyr::collect) {
  parts_filtered <- input_func("part") %>%
    filter(
      p_brand == "Brand#23",
      p_container == "MED BOX"
    )

  joined <- input_func("lineitem") %>%
    inner_join(parts_filtered, by = c("l_partkey" = "p_partkey"))

  quantity_by_part <- joined %>%
    group_by(l_partkey) %>%
    summarise(quantity_threshold = 0.2 * mean(l_quantity))

  joined %>%
    inner_join(quantity_by_part, by = "l_partkey") %>%
    filter(l_quantity < quantity_threshold) %>%
    summarise(avg_yearly = sum(l_extendedprice) / 7.0) %>%
    collect_func()
}

File: ibis_tpc/h17.py

def tpc_h17(con, BRAND='Brand#23', CONTAINER='MED BOX'):
    '''Small-Quantity-Order Revenue Query (Q17)

    This query determines how much average yearly revenue would be lost if
    orders were no longer filled for small quantities of certain parts. This
    may reduce overhead expenses by concentrating sales on larger shipments.'''

    lineitem = con.table('lineitem')
    part = con.table('part')

    q = lineitem.join(part, part.P_PARTKEY == lineitem.L_PARTKEY)
    q = q.materialize()

    innerq = lineitem
    innerq = innerq[
        lineitem.L_EXTENDEDPRICE.name('extprice'),
        lineitem.L_QUANTITY
    ]
    innerq = innerq.filter([lineitem.L_PARTKEY == q.P_PARTKEY])

    q = q.filter([
        q.P_BRAND == BRAND,
        q.P_CONTAINER == CONTAINER,
        q.L_QUANTITY < (0.2*innerq.L_QUANTITY.mean())
    ])
    q = q.aggregate(avg_yearly=innerq.extprice.sum()/7.0)
    return q
